lib/experimental/colortext.nim

```diff
@@ -0,0 +1,781 @@
+  ## Append one or more colored runes to the text
+  text.runes.add rune
+
+iterator items*(text: ColText): ColRune =
+  ## Iterate over all runes left to right
+  for item in text.runes:
+    yield item
+
+iterator ritems*(text: ColText): ColRune =
+  ## Iterate over all runes right to left
+  var idx = text.runes.high
+  while 0 <= idx:
+    yield text.runes[idx]
+    dec idx
+
+func `==`*(s1, s2: ColStyle): bool =
+  ## COmpare styles for equality - foreground, background and style
+  s1.fg == s2.fg and s1.bg == s2.bg and s1.style == s2.style
+
+func contains*(ps: ColStyle, s: Style): bool =
+  ## Check if coloring contains specific style option
+  ps.style.contains(s)
+
+func term*(bg: BackgroundColor): TermColorBg =
+  ## Convert `std/terminal.BackgroundColor` to terminal background color
+  if bg != bgDefault:
+    result = TermColorBg(bg.uint8 - low(BackgroundColor).uint8)
+
+func term*(bg: ForegroundColor): TermColorFg =
+  ## Convert `std/terminal.ForegroundColor` to terminal foreground color
+  if bg != fgDefault:
+    result = TermColorFg(bg.uint8 - low(ForegroundColor).uint8)
+
+func `==`*(f1: TermColorFg, f2: ForegroundColor): bool =
+  ## Compare for equality with `std/terminal` color
+  term(f2) == f1
+
+func `==`*(f1: TermColorBg, f2: BackgroundColor): bool =
+  ## Compare for equality with `std/terminal` color
+  term(f2) == f1
+
+
+func initColStyle*(
+    fg: ForegroundColor = fgDefault,
+    bg: BackgroundColor = bgDefault,
+    style: set[Style] = {}
+  ): ColStyle =
+  ## Initialize color style with given options
+  ColStyle(fg: fg.term, bg: bg.term, style: style)
+
+func colStyle*(fg: ForegroundColor): ColStyle =
+  ## Convert to style with only foreground option set
+  ColStyle(fg: fg.term)
+
+func colStyle*(bg: BackgroundColor): ColStyle =
+  ## Convert to style with only background option set
+  ColStyle(bg: bg.term)
+
+func colStyle*(fg: TermColorFg): ColStyle =
+  ## Convert to stylewith only foreground color set
+  ColStyle(fg: fg)
+
+func colStyle*(bg: TermColorBg): ColStyle =
+  ## Convert to style with only background color set
+  ColStyle(bg: bg)
+
+func colStyle*(style: set[Style]): ColStyle =
+  ## Convert to style with given set of styling options
+  ColStyle(style: style)
+
+func colStyle*(style: Style): ColStyle =
+  ## Convert tot style with single styling option set
+  ColStyle(style: {style})
+
+func colStyle*(style: ColStyle): ColStyle =
+  ## Passthrough proc, used to simplify mapping of the `ColModifier`
+  ## typeclass to the `ColStyle` object.
+  style
+
+func default*(rune: typedesc[ColRune]): ColRune =
+  ## Default value of the colored rune
+  ColRune(style: initColStyle())
+
+func default*(style: typedesc[ColStyle]): ColStyle =
+  ## Default value of the color style
+  ColStyle(fg: fgDefault.term(), bg: bgDefault.term())
+
+func uc*(s: static[string]): Rune =
+  ## Create single unicode rune from string literal - `uc"⮰"`
+  runeAt(s, 0)
+
+func isDefault*(col: TermColorFg | TermColorBg): bool =
```

Comment:
Right about here is where you confirmed my thinking that we could cut down a lot of repetition by using type classes.